cmake/tests.cmake

```diff
@@ -48,3 +48,51 @@ int main(int argc, char* argv[]) {\n\
   target_link_libraries(GTest_Main PUBLIC GTest::GTest)
   add_library(GTest::Main ALIAS GTest_Main)
 endif()
+
+if(NOT TARGET GTest::QtCoreMain AND TARGET Qt5::Core)
+  set(FILE_PATH ${CMAKE_BINARY_DIR}/test_qtcore_main.cpp)
+  if(NOT EXISTS ${FILE_PATH})
+    set(GTEST_MAIN_CPP "\
+#include <cstdio>\n\
+#include <gtest/gtest.h>\n\
+#include <QCoreApplication>\n\
+\n\
+int main(int argc, char* argv[]) {\n\
+  QCoreApplication app{argc, argv}\;\n\
+  printf(\"Running main() from %s\\n\", __FILE__)\;\n\
+  ::testing::InitGoogleTest(&argc, argv)\;\n\
+  return RUN_ALL_TESTS()\;\n\
+}\n\
+")
+
+    file(WRITE ${FILE_PATH} ${GTEST_MAIN_CPP})
+  endif()
+
+  add_library(GTest_QtCoreMain STATIC EXCLUDE_FROM_ALL ${FILE_PATH})
+  target_link_libraries(GTest_QtCoreMain PUBLIC GTest::GTest Qt5::Core)
+  add_library(GTest::QtCoreMain ALIAS GTest_QtCoreMain)
+endif()
+
+if(NOT TARGET GTest::QtGuiMain AND TARGET Qt5::Widgets)
+  set(FILE_PATH ${CMAKE_BINARY_DIR}/test_qtgui_main.cpp)
+  if(NOT EXISTS ${FILE_PATH})
+    set(GTEST_MAIN_CPP "\
+#include <cstdio>\n\
+#include <gtest/gtest.h>\n\
+#include <QApplication>\n\
+\n\
+int main(int argc, char* argv[]) {\n\
+  QApplication app{argc, argv}\;\n\
+  printf(\"Running main() from %s\\n\", __FILE__)\;\n\
+  ::testing::InitGoogleTest(&argc, argv)\;\n\
+  return RUN_ALL_TESTS()\;\n\
+}\n\
+")
+
+    file(WRITE ${FILE_PATH} ${GTEST_MAIN_CPP})
+  endif()
+
+  add_library(GTest_QtGuiMain STATIC EXCLUDE_FROM_ALL ${FILE_PATH})
+  target_link_libraries(GTest_QtGuiMain PUBLIC GTest::GTest Qt5::Widgets)
```

Comment:
If I see this correctly the only difference between `QtCoreMain` and `QtGuiMain` is that the latter links against `Qt5::Widgets`? Interestingly the Tutorials-Test, which instantiates a UI, only requires `QtCoreMain`, so I wonder in which case we'd need `QtGuiMain`?

Reply:
It depends on whether you need a QApplication or QCoreApplication object. The former provides everything the latter does and additionally things like keeping track of windows, handling user input (keyboard and mouse events), clipboard support, handling of screens, etc.

It's not needed to instantiate Widgets. But as soon as you want some of the mentioned features you need a QApplication instance over a QCoreApplication instance.